### PR TITLE
ClientModel: Update BinaryContent.Create(Stream) implementation based on PR feedback

### PR DESCRIPTION
**Pull request by annelo-msft**
Addresses https://github.com/Azure/azure-sdk-for-net/pull/42696#discussion_r1529161913

We had removed a CanSeek check from the BinaryContent.TryGetLength method based on the assumption that CanSeek has the same value for the full lifetime of the Stream.

One thing to note is that, if the Stream changes from seekable to not-seekable, than calls to WriteTo will throw as well.  This just moves the problem to a different place, since most HttpContent implementations first call TryGetLength and then call WriteTo.  We should decide as part of this PR review whether we want to throw a different exception from the WriteTo methods or let the Stream implementation's exception be what gets thrown.

### CODE REVIEW

File: sdk/core/System.ClientModel/src/Message/BinaryContent.cs
Comment (christothes) [inline sdk/core/System.ClientModel/src/Message/BinaryContent.cs:226]: Can you remind me why we don't provide the length directly as opposed to giving the length remaining after the current stream position?

File: sdk/core/System.ClientModel/src/Message/BinaryContent.cs
Comment (christothes) [inline sdk/core/System.ClientModel/src/Message/BinaryContent.cs:1]: > One thing to note is that, if the Stream changes from seekable to not-seekable, than calls to WriteTo will throw as well. This just moves the problem to a different place, since most HttpContent implementations first call TryGetLength and then call WriteTo. We should decide as part of this PR review whether we want to throw a different exception from the WriteTo methods or let the Stream implementation's exception be what gets thrown.

If failure is inevitable, it feels like we should fail as early as possible and let the `ObjectDisposedException` bubble up for any operation where the stream would also throw.

File: sdk/core/System.ClientModel/src/Message/BinaryContent.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/src/Message/BinaryContent.cs:226]: I'm not 100% confident of the original author's intent, but I had assumed it was in case someone was writing to a stream starting from some non-zero position.  I've been poking around the BCL HttpContent code, and it looks like things like MultipartContent compute their Content-Length by calling this method on their "parts" and then adding in the length of the content segments that they add when serializing to a stream.  So, it could be possible that the full stream length (_stream.Length) represents more than the length of this instance's length.

File: sdk/core/System.ClientModel/src/Message/BinaryContent.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/src/Message/BinaryContent.cs:1]: Yeah, I was feeling that way too as I was looking at this today.  If that's the case and we agree as a Core team about that, I think the next question is whether we would make a breaking change to Azure.Core to achieve the same behavior across the two libraries.  I think there is a good argument for consistency here -- it seems very unlikely that someone was calling TryGetLength without also calling WriteTo, so it would just move the same exception a little early in the call sequence.

File: sdk/core/Azure.Core/src/RequestContent.cs
Comment (jsquire) [inline sdk/core/Azure.Core/src/RequestContent.cs:179]:
```suggestion
                    throw new ArgumentException("Stream must be seekable.", nameof(stream));
```

Comment (jsquire) [reply sdk/core/Azure.Core/src/RequestContent.cs:179]: Though, we may find this undesirable because it's an existing error... but the lack of proper sentence makes me sad.

**Review by jsquire — APPROVED** [review]

**Comment by annelo-msft** [comment]
Given the desire to bubble up the `ObjectDisposedException` early, closing this PR in favor of keeping the existing implementation.

**Pull request closed without merging**
